lookout/style/format/rules.py

```diff
@@ -782,66 +786,61 @@ def _prune_branches_top_down_greedy(
         new_rules = []
-        intervals = {}
-        attrs = defaultdict(set)
-        for branch, _, _ in rules:
-            for rule_attr in branch:
-                attrs[rule_attr.feature].add(rule_attr.threshold)
-        for key, vals in attrs.items():
-            attrs[key] = numpy.array(sorted(vals))
-            intervals[key] = [defaultdict(int) for _ in range(len(vals) + 1)]
-        searchsorted = numpy.searchsorted
-        for x, y in zip(X.toarray(), Y):
-            for attr, val in enumerate(x):
-                interval = intervals.get(attr)
-                if interval is not None:
-                    interval[searchsorted(attrs[attr], val)][y] += 1
-        for key, vals in attrs.items():
-            attrs[key] = {v: i for i, v in enumerate(vals)}
-        for vals in intervals.values():
-            for vec in vals:
-                vec[-1] = sum(vec.values())
+        if cls._log.isEnabledFor(logging.INFO):
+            rules = tqdm(rules)
+        x_array = X.toarray()
+        not_cs = {}
         for rule, stats, artificial in rules:
+            if artificial:
+                new_rules.append(rule)
+                continue
             c = stats.cls
-            new_verbs = []
+            not_c = not_cs.setdefault(c, Y != c)
+            triggered = []
             for feature, cmp, thr in rule:
-                table = numpy.zeros((2, 2), dtype=numpy.int32)
-                for i, interval in enumerate(intervals[feature]):
-                    row = int((i <= attrs[feature][thr]) == cmp)
-                    num_same_cls = interval[c]
-                    table[row, 0] += num_same_cls
-                    table[row, 1] += interval[-1] - num_same_cls
-                if prune_uncertain:
-                    if confidence(table[0, 0] + table[1, 0], table[0, 1] + table[1, 1]) \
-                            >= confidence(table[0, 0], table[0, 1]):
-                        continue
-                _, p = fisher_exact(table)
-                if p < 0.01:
-                    new_verbs.append(RuleAttribute(feature, cmp, thr))
-            if new_verbs:
-                new_rules.append(Rule(attrs=tuple(new_verbs), stats=stats, artificial=artificial))
+                if cmp:
+                    triggered.append(numpy.nonzero((x_array[:, feature] > thr) & not_c)[0])
+                else:
+                    triggered.append(numpy.nonzero((x_array[:, feature] <= thr) & not_c)[0])
+            graph = Graph()
+            graph.add_vertices(len(rule))
+            for x, tx in enumerate(triggered):
+                for y, ty in enumerate(triggered[x + 1:]):
+                    y += x + 1
+                    sim = len(set(tx).intersection(set(ty))) / len(set(tx).union(set(ty)))
+                    if sim > sim_threshold:
+                        graph.add_edge(x, y)
+            erased = set()
+            retained = set()
+            cliques = sorted(graph.cliques(2), key=lambda c: -len(c))
```

Comment:
Up to you, but I prefer:
```suggestion
            cliques = sorted(graph.cliques(2), key=len, reverse=True)
```